core/src/main/java/com/yahoo/oak/EntrySet.java

```diff
@@ -777,28 +609,40 @@ void writeValueFinish(ThreadContext ctx) {
      * is responsible to call it inside the publish/unpublish scope.
      */
     boolean deleteValueFinish(ThreadContext ctx) {
-        final int version = ctx.value.getVersion();
-        if (version <= INVALID_VERSION) { // version is marked deleted
-            return false;
+        if (valueRC.isReferenceDeleted(ctx.value)) {
+            return false; // value reference in the slice is marked deleted
         }
 
         assert ctx.valueState == ValueState.DELETED_NOT_FINALIZED;
 
         int indIdx = entryIdx2intIdx(ctx.entryIndex);
-        // Scenario: this value space is allocated once again and assigned into the same entry,
-        // while this thread is sleeping. So later a valid value reference is CASed to invalid.
-        // In order to not allow this scenario happen we must release the
-        // value's off-heap slice to memory manager only after deleteValueFinish is done.
-        casEntriesArrayLong(indIdx, OFFSET.VALUE_REFERENCE,
-                ctx.value.getReference(), ReferenceCodec.INVALID_REFERENCE);
-        if (casEntriesArrayInt(indIdx, OFFSET.VALUE_VERSION, version, -version)) {
+
+        // Value's reference codec prepares the reference to be used after value is deleted
+        long expectedReference = ctx.value.getReference();
+        if (casEntriesArrayLong(indIdx, OFFSET.VALUE_REFERENCE, expectedReference,
+            valueRC.alterForDelete(expectedReference))) {
+            assert ((ReferenceCodecMM) valueRC).isReferenceConsistent(getValueReference(ctx.entryIndex));
             numOfEntries.getAndDecrement();
-            memoryManager.release(ctx.value);
+            valuesMemoryManager.release(ctx.value);
             ctx.value.invalidate();
             ctx.valueState = ValueState.DELETED;
+
             return true;
         }
+        // CAS wasn't successful, someone else set the value reference as deleted and
+        // maybe value was allocated again. Let the context's value to be updated
+        readValue(ctx);
         return false;
+
+        // Scenario:
+        // 1. The value's slice is marked as deleted off-heap and the thread that started
+        //    deleteValueFinish falls asleep.
+        // 2. This value's slice space is allocated once again and assigned into the same entry.
+        // 3. A valid value reference is CASed to invalid.
+        //
+        // This is ABA problem and resolved via always changing deleted variation of the reference
+        // Also value's off-heap slice is released to memory manager only after deleteValueFinish
+        // is done.
```

Comment:
The solution is manifested in the fact that in the lines above 'valuesMemoryManager.release(ctx.value);' is invoked only after deleteValueFinish succeed. And not before the CAS and not just after off-heap header deleted bit is set (via CAS as well). Am I answering the question?

Reply:
Yes. I think this is also what we discussed online.
I think this comment should be placed right before the CAS.
Maybe even separate `valuesMemoryManager.alterReferenceForDelete(expectedReference)` to its own line and place this comment right above it, so it will be perfectly clear what it refers to.